src/audio/google/google_rtc_audio_processing.c

```diff
@@ -822,23 +821,60 @@ static int google_rtc_audio_processing_process(struct processing_module *mod,
 	/* 32float: de-interlace ref buffer, convert it to float, skip channels if > Max
 	 * 16int: linearize buffer, skip channels if > Max
 	 */
+	/* Reduce cycle waste by streamlining the inner loop,
+	 * converting from array indexing to pointer arithmetic,
+	 * and putting data copy verification outside the loop.
+	 */
 	buffer_offset = 0;
-	for (int i = 0; i < cd->num_frames; i++) {
-		for (channel = 0; channel < cd->num_aec_reference_channels; ++channel) {
-#if CONFIG_COMP_GOOGLE_RTC_USE_32_BIT_FLOAT_API
-			cd->aec_reference_buffer_ptrs[channel][i] =
-					convert_int16_to_float(ref[channel]);
-#else /* CONFIG_COMP_GOOGLE_RTC_USE_32_BIT_FLOAT_API */
-			cd->aec_reference_buffer[buffer_offset++] = ref[channel];
-#endif /* CONFIG_COMP_GOOGLE_RTC_USE_32_BIT_FLOAT_API */
+	const int16_t *ref_end = ref + cd->num_frames * cd->num_aec_reference_channels;
+
+	if ((void *)ref_end >= (void *)ref_buf_end)
+		ref_end = (void *)ref_buf_start;
 
+#if CONFIG_COMP_GOOGLE_RTC_USE_32_BIT_FLOAT_API
+	float **ref_ptr = cd->aec_reference_buffer_ptrs;
+	int s_chan;
+	int i;
+
+	/*  Loop over frames and channels, converting data from int16 to float */
+	for (i = 0; i < cd->num_frames; ++i) {
+		for (s_chan = 0; s_chan < cd->num_aec_reference_channels; ++s_chan) {
+		/*  Check that ref is within the valid range of the ref_buf buffer */
+			if (ref && (void *)ref >= (void *)ref_buf_start &&
+			    (void *)ref < (void *)ref_buf_end)
```

Comment:
no need to check `ref` for `NULL` at all. You anyway check, that it's in a range